### PR TITLE
[SPARK-47907][SQL] Put bang under a config

### PR DESCRIPTION
**Pull request by srielau**
### What changes were proposed in this pull request?

Recently a bug was fixed in the lexer which equated ! (bang) with NOT. The original bug allowed for unintended syntax such as:
CREATE TABLE IF ! EXISTS TY(c1 INT ! NULL)
and also, slightly less iffy: c1 ! IN (1, 2)

It turns out this undocumented "feature" can be found in the wild with examples of `! LIKE`, `! BETWEEN` and `! IN`.
So we need to provide a safety valve by allowing users to re-enable the bug, at least for a while.

### Does this PR introduce _any_ user-facing change?

Yes (The original bugfix does, here we make it offical)

### How was this patch tested?

Added automated  testcases for both modes.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by srielau** [comment]
@cloud-fan @gengliangwang This is ready for review.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala [hunk 1]
```diff
@@ -365,6 +367,8 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper with Logging {
     val cols = Option(ctx.identifierList()).map(visitIdentifierList).getOrElse(Nil)
     val partitionKeys = Option(ctx.partitionSpec).map(visitPartitionSpec).getOrElse(Map.empty)
 
+    blockBang(ctx.errorCapturingNot())
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:370]: does it work if we override `def visitErrorCapturingNot` and put the check there?

File: sql/core/src/test/resources/sql-tests/analyzer-results/predicate-functions.sql.out [hunk 2]
```diff
@@ -450,3 +450,18 @@ Project [NOT between(to_timestamp(2022-12-26 00:00:01, None, TimestampType, Some
...
 -- !query analysis
 [Analyzer test output redacted due to nondeterminism]
+
+
+-- !query
+set spark.sql.legacy.bangEqualsNot=true
+select 1 ! between 0 and 2
+-- !query analysis
+java.lang.IllegalArgumentException
+spark.sql.legacy.bangEqualsNot should be boolean, but was true
```
Comment (cloud-fan) [inline sql/core/src/test/resources/sql-tests/analyzer-results/predicate-functions.sql.out:460]: this looks weird...

File: common/utils/src/main/resources/error/error-conditions.json [hunk 3]
```diff
@@ -3798,6 +3798,21 @@
...
     "sqlState" : "22003"
   },
+  "SYNTAX_DISCONTINUED" : {
+    "message" : [
+      "Support of the clause or keyword: <clause> has been discontinued in this context."
+    ],
+    "subClass" : {
+      "BANG_EQUALS_NOT" : {
+        "message" : [
+          "The '!' keyword is supported as a prefix operator in a logical operation only.",
```
Comment (cloud-fan) [inline common/utils/src/main/resources/error/error-conditions.json:3808]: what is logical operation?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (srielau) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:370]: I tried that. But I need to invoke visitErrorCapturingNot explicitly which makes it worse.

File: sql/core/src/test/resources/sql-tests/analyzer-results/predicate-functions.sql.out (hunk 2 above)
Comment (srielau) [reply sql/core/src/test/resources/sql-tests/analyzer-results/predicate-functions.sql.out:460]: Oh, missing semicolon....

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala (hunk 1 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:370]: ah I see

**Review by cloud-fan — APPROVED** [review]

File: common/utils/src/main/resources/error/error-conditions.json (hunk 3 above)
Comment (srielau) [reply common/utils/src/main/resources/error/error-conditions.json:3808]: NOT, AND, OR (?) I can rephrase, "Only supported for the prefix operator "NOT"?

Comment (srielau) [reply common/utils/src/main/resources/error/error-conditions.json]: Rephrased

File: sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql [hunk 4]
```diff
@@ -82,3 +82,8 @@ select 2.0 not between '1.0' and '3.0';
 select 'b' not between 'a' and 'c';
 select to_timestamp('2022-12-26 00:00:01') not between to_date('2022-03-01') and to_date('2022-12-31');
 select rand(123) not between 0.1 AND 0.2;
+
+-- Sanity test for legacy flag equating ! with NOT
+set spark.sql.legacy.bangEqualsNot=true;
+select 1 ! between 0 and 2;
```
Comment (gengliangwang) [inline sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql:88]: nit: let's add some tests cases that bang equals to not.

Comment (srielau) [reply sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql:88]: Not sure why we need to tickle the codepath more than once...

Comment (gengliangwang) [reply sql/core/src/test/resources/sql-tests/inputs/predicate-functions.sql:88]: It's ok. I can have a followup to add more tests.

**Review by gengliangwang — APPROVED** [review]

**Comment by gengliangwang** [comment]
Thanks, merging to master

**Comment by gengliangwang** [comment]
Thanks, merging to master.

**Pull request closed without merging**